src/env_universal_common.rs

```diff
@@ -154,90 +131,76 @@ impl EnvUniversal {
 
     /// Initialize this uvars for the default path.
     /// This should be called at most once on any given instance.
-    pub fn initialize(&mut self, callbacks: &mut CallbackDataList) {
+    pub fn initialize(&mut self) -> Option<CallbackDataList> {
         // Set do_flock to false immediately if the default variable path is on a remote filesystem.
         // See #7968.
         if path_get_config_remoteness() == DirRemoteness::remote {
-            self.do_flock = false;
+            LOCK_UVAR_FILE.store(false);
         }
-        self.initialize_at_path(callbacks, default_vars_path());
+        self.initialize_at_path(default_vars_path())
     }
 
     /// Initialize a this uvars for a given path.
     /// This is exposed for testing only.
-    pub fn initialize_at_path(&mut self, callbacks: &mut CallbackDataList, path: WString) {
+    pub fn initialize_at_path(&mut self, path: WString) -> Option<CallbackDataList> {
         if path.is_empty() {
-            return;
+            return None;
         }
         assert!(!self.initialized(), "Already initialized");
         self.vars_path = path;
 
-        if self.load_from_path(callbacks) {
-            // Successfully loaded from our normal path.
-        }
+        self.load_from_path()
     }
 
     /// Reads and writes variables at the correct path. Returns true if modified variables were
     /// written.
-    pub fn sync(&mut self, callbacks: &mut CallbackDataList) -> bool {
+    pub fn sync(&mut self) -> (bool, Option<CallbackDataList>) {
```

Comment:
we should generally prefer (named) structs over tuples.
This one is probably okay to leave but we should try to define a struct for the return value of `load_from_file()`, since things like `u64` don't give a lot of information

```diff
-    /// That is:
-    /// 1. by how much `self.export_generation` should be incremented
-    /// 2. the new variable table (to replace `self.vars`)
-    /// 3. callbacks
-    fn load_from_file(
-        &self,
-        file: &File,
-    ) -> std::io::Result<Option<(u64, VarTable, CallbackDataList)>> {
+struct UniversalReadUpdate {
+    export_generation_increment: u64,
+    new_vars: VarTable,
+    callbacks: CallbackDataList,
+}
+    fn load_From_file(&self, file: &File) -> std::io::Result<Option<UniversalReadUpdate>> {
```

not sure, I didn't try that